Docs / naming updates for pkg/autodiscovery

djmitche opened the pull request:
### What does this PR do?

This adds/updates docs in pkg/autodiscovery, and also renames things (mostly to make them private and/or clearer).  It should have no functional impact.

### Motivation

Helping the next person learning how this stuff works :)

### Additional Notes

As with #10378, please verify my statements in the docs -- I've guessed in some places!  There's also a bonus "XXX" in there that remains unanswered :)
 

### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).

Note: Adding GitHub labels is only possible for contributors with write access.
-->

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] The appropriate `team/..` label has been applied, if known.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

djmitche committed:
update README with current structure

djmitche committed:
update Autoconf doc comment, removing old info

djmitche committed:
Remove outdated comment

This comment was added to document a code change in
b93e1f4e30a436a50c6314e42605864d1bf054e2.  The code has since been
reverted.

djmitche committed:
Describe what 'loaded configs' means

djmitche committed:
make pkg/autodiscovery.TemplateCache private

djmitche committed:
doc ConfigProvider methods

djmitche committed:
integration.Config docs + drop unused method

The Type() method is unused, and did not distinguish template,
non-template, and service configs.

djmitche committed:
make ProviderCache private, doc it, and use clearer field names

ahmed-mez commented on pkg/autodiscovery/README.md:
```diff
@@ -1,15 +1,55 @@
+               │ Workload Meta  │
+               └──┬──────────┬──┘
+  Static          │          │
+  Files──┐        │          │
+         │        │          │
+    ┌────▼────────▼──┐    ┌──▼─────────────┐
+    │Config Providers│    │   Listeners    │
+    └──┬─────┬───────┘    └────────┬───────┘
```
Nit: there is another config source, the `clusterchecks` and `endpointschecks` config provider consumes configs from the datadog cluster agent.

ahmed-mez commented on pkg/autodiscovery/README.md:
```diff
@@ -1,15 +1,55 @@
+Entities that contain their own configuration are reconciled using an AD identifier unique to that entity.
+For example, a new container might be deteted first by a listener, creating a new serivce with an AD identifier containing its SHA.
+Soon after, the relevant config provider detects the container, extracts configuration from its labels, and creates an `integration.Config` containing the same AD identifier.
+The reconciliation process combines the service and the Config, resolving the template, and schedules the resolved config.
```
We can mention [template variables](https://docs.datadoghq.com/agent/faq/template_variables/) here to clarify what we mean by `resolving the template`.

ahmed-mez reviewed (approved):
Thank you!

djmitche committed:
include cluster/endpoint checks and kube_*

djmitche committed:
mention template variables

ahmed-mez reviewed (approved):
Looks like the CI is failing because of a typo in `config.go`. LGMT otherwise 👍

djmitche commented on pkg/autodiscovery/integration/config.go:
```diff
@@ -37,27 +37,78 @@ const (
+	// Entity is the entity ID of the service, set only for service configs.
+	Entity string `json:"-"` // (include in digest: true)
+
+	// TaggerEntity is the tagger entity ID
+	// XXX how is this different from Entity?
+	TaggerEntity string `json:"-"` // (include in digest: false)
```
Even this part? :)

ahmed-mez replied:
IIRC we were using the same entity format for AD and Tagger (like 2 and half years ago) the format was `<container runtime>://<id>` which is still the same format for `EntityID`. And then we started supporting APM tagging; tracers parse their container cgroups and send their container ID to the trace agent, the latter uses that container ID to query the tagger and enrich traces with container tags. The tracers don't have access to container runtime sockets and they don't know their container runtime, so they just send the <id> part, to make this work generically for any container runtime, we introduced `TaggerEntity` with the format `container_id://<id>` to simplify APM container tagging, that's why every tagger client should respect this format now.

djmitche replied:
OK, I'll see if I can add some documentation about that.  To verify I have it right, there are currently two entity IDs formats:

 * WLM Entity:
   * `"container://<sha>"` for containers
   * `"garden_container://<sha>"` for garden containers
   * `"kubernetes_pod_uid://<uid>"` for kubernetes pods
   * `"ecs_task://<task-id>"` for ECS tasks
 * Tagger Entity:
   * `"container_id://<sha>"` for containers
   * `"container_id://<sha>"` for garden containers
   * `"kubernetes_pod://<uid>"` for kubernetes pods
   * `"ecs_task://<task-id>"` for ECS tasks
  
That seems massively unfortunate -- so close to matching, yet not! -- but what can you do?  If it's correct, I'll add it to the docs.

djmitche replied:
I got that slightly wrong, I think -- WLM is careful to never use URL-like syntax, but AD does.  It seems that AD basically uses "Entity" as a name for a service, and that might be a good rename.  I'll update this PR to match.

djmitche committed:
fix typo

djmitche committed:
Rename AD 'entity' to 'service ID'

This field is, in fact, a unique identifier for the AD service and is
not to be confused with entities as defined by either the tagger or the
workloadmeta service.  Renaming the helps avoid that confusion.

djmitche committed:
don't use containers.BuildEntityName for things that aren't containers

djmitche committed:
clarify Config.Entity vs. Config.TaggerEntity

djmitche committed:
Rename integration.Config.Entity to ServerID

This better reflects its meaning, as it does not correspond to either a
WLM entity or a Tagger Entity.

djmitche committed:
do not treat serviceIDs as entity IDs

djmitche commented:
Apologies for the force-push (I applied the fixup early in the commit stack).  Please have a look at the commits from "fix typo" on.  I did two fairly wide-reaching renames, but they shouldn't have any affect on program logic.

djmitche committed:
fix missed rename

djmitche commented on pkg/autodiscovery/autoconfig.go:
```diff
@@ -662,15 +656,15 @@ func (ac *AutoConfig) processNewService(ctx context.Context, svc listeners.Servi
 		{
 			LogsConfig:      integration.Data{},
-			Entity:          svc.GetEntity(),
+			ServiceID:       svc.GetServiceID(),
```
Thanks for looking!  Calling these "entities" is what initially confused me and led to this change.  The things this value labels are of type `pkg/autodiscovery/listeners.Servce`, so I think calling it a "Service ID" makes more sense than "Entity ID" (since there is no struct named "Entity" in the AD package).   WDYT?

djmitche commented on pkg/autodiscovery/integration/config.go:
```diff
@@ -37,27 +37,78 @@ const (
+//
+// The Data fields contain YAML data except when config.Provider is
+// names.Container or names.Kubernetes, in which case the configuration is in
+// JSON.
```
^^ I'd still love an answer to this cc @ahmed-mez :)

ahmed-mez replied:
Correct! Checks are defined in JSON on docker labels and pod annotations.

davidor commented on pkg/autodiscovery/README.md:
````diff
@@ -1,15 +1,92 @@
+       │         ┌─────────────┐   │
+       └────────►│metascheduler│◄──┘
+                 └─────────────┘
+```
````
Question: Are there services that go directly from the listener to the metascheduler without being reconciled with a template?

ahmed-mez replied:
Just to add context: this a technical debt to make the logs agent learn about all containers and support `logs_config.container_collect_all`. It should be replaced by subscribing directly to workloadmeta in the coming releases 🤞

djmitche replied:
That's what I'm working on -- but aspirational docs are an anti-pattern, so let's doc it as it is.

ahmed-mez replied:
Of course, we should document it as it is 👌

ogaca-dd reviewed (approved):
LGTM. Great PR!

djmitche commented on pkg/autodiscovery/README.md:
```diff
@@ -1,15 +1,92 @@
+
+The [listeners](https://pkg.go.dev/github.com/DataDog/datadog-agent/pkg/autodiscovery/listeners) monitor entities, known as "services" in this package, such as pods, containers, or tasks.
+
+Each service has two entity identifiers: the AD service ID (from `svc.GetServiceID()`) and the Tagger entity (`svc.GetTaggerEntity()`).
```
Yes, I'd like to do that in a followup.

djmitche commented on pkg/tagger/collectors/workloadmeta_extract.go:
```diff
@@ -588,8 +588,9 @@ func buildTaggerEntityID(entityID workloadmeta.EntityID) string {
-		return containers.BuildEntityName(string(entityID.Kind), entityID.ID)
+		log.Errorf("can't recognize entity %q with kind %q; trying %s://%s as tagger entity",
+			entityID.ID, entityID.Kind, entityID.ID, entityID.Kind)
+		return fmt.Sprintf("%s://%s", string(entityID.Kind), entityID.ID)
```
In this case it's a serviceID and not an entityID.  This would be clearer if those two things returned different types :)

djmitche commented on pkg/autodiscovery/common_test.go:
```diff
@@ -24,12 +24,12 @@ type dummyService struct {
 }
 
-// GetEntity returns the service entity name
+// GetServiceID returns the service entity name
```
Ah, this is a rename done automatically by gopls.  It probably got confused :)

djmitche committed:
fix incorrect doc comments in tests

djmitche committed:
fix README typos

djmitche commented on pkg/logs/scheduler/scheduler.go:
```diff
@@ -256,6 +256,16 @@ func (s *Scheduler) parseEntity(entity string) (string, string, error) {
 	return components[0], components[1], nil
 }
 
+// parseServiceID breaks down an AD service ID, assuming it is formatted
```
It's still called, anyway.  Maybe it will be removed in the followup that strongly-types the IDs?
